Add cache expire feature

ketchoop opened the pull request:
Введена проверка кэшей по времени жизни. Включается с помощью перменной окружения NPM_CACHE_EXPIRE_TIME. Время жизни кэша указывается в днях.

ketchoop committed:
Add cache expire feature

shuhrat commented on npm:
```diff
@@ -38,6 +38,24 @@ fi
+  echo "Looking for expired cache"
+
+  NPM_CACHE_DIR="$HOME/.cache/npm-inject"
+  EXPIRED_DIRS=$(find $NPM_CACHE_DIR -maxdepth 1 -mtime +$NPM_CACHE_EXPIRE_TIME ! -path $NPM_CACHE_DIR -type d | tr \\n ' ')
```
думаю тут можно обойтись без `! path` и `| tr \\n ' '`

ketchoop replied:
По поводу tr -- согласен, забыл, что for воспринимает такие символы в качестве разделителя и не нужно менять IFS:smile:. А вот с path интереснее, разве он не будет без него выдавать в некоторых случаях и текущую папку (имею ввиду './')? Она не удалится, но errlog bash'а с удалением этой папки видеть неприятно, а глушить их в /dev/null я думаю не стоит.

shuhrat commented on npm:
```diff
@@ -38,6 +38,24 @@ fi
 echo "==> Running npm wrapper script with node_modules caching"
 
+#Если существует переменная времени жизни кэша, то ищем просроченные кэши
+#Время жизни указывается в днях в переменно $NPM_CACHE_EXPIRE_TIME
```
`NPM_CACHE_EXPIRE_TIME` → `NPM_CACHE_EXPIRE_DAYS`

shuhrat commented on npm:
```diff
@@ -38,6 +38,24 @@ fi
+  if [[ -n $EXPIRED_DIRS ]]; then
+    for EXPIRED_DIR in $EXPIRED_DIRS; do
+      echo "Clear expired cache: $EXPIRED_DIR"
+      rm -r $EXPIRED_DIR
```
`rm -r "$EXPIRED_DIR"`

shuhrat commented on npm:
```diff
@@ -86,3 +104,4 @@ else
   echo "Linking $CACHE_DIR/node_modules to $NODE_MODULES"
   ln -s "$CACHE_DIR/node_modules" "$NODE_MODULES"
 fi
+
```
пустая строка

shuhrat commented on npm:
```diff
@@ -38,6 +38,24 @@ fi
 
+#Если существует переменная времени жизни кэша, то ищем просроченные кэши
+#Время жизни указывается в днях в переменно $NPM_CACHE_EXPIRE_TIME
+if [[ -n ${NPM_CACHE_EXPIRE_TIME} ]]; then
```
Может унести это блок в конец скрипта?

seth2810 replied:
@shuhrat предположу, что это делается для того, чтобы снести кэш перед установкой, если в конец поместить совсем другой эффект будет

ketchoop replied:
@seth2810 именно так. Если унести в конец, то будет неприятный эффект в виде удаления единственного пролинкованного и устаревшего кэша. Придётся запускать два раза.

seth2810 commented on npm:
```diff
@@ -38,6 +38,24 @@ fi
+if [[ -n ${NPM_CACHE_EXPIRE_TIME} ]]; then
+  echo "Looking for expired cache"
+
+  NPM_CACHE_DIR="$HOME/.cache/npm-inject"
```
CACHE_DIR="$HOME/.cache/npm-inject"
PACKAGES_CACHE_DIR="$CACHE_DIR/$PACKAGES_HASH" - https://github.com/shuhrat/npm-inject/blob/master/npm#L52

чтобы не повторяться

ketchoop committed:
Minor fixes according to 'change requests' 

Change "NPM_CACHE_EXPIRE_TIME" to "NPM_CACHE_EXPIRE_DAYS"
Delete redundant "tr" command call
Delete blank line in the end of the code.

ketchoop committed:
Add quotes and "force" flag to rm expired command

seth2810 commented:
https://github.com/shuhrat/npm-inject/pull/1#discussion_r91269054 @ket4yii ты на это забил?

ketchoop commented:
@seth2810 нет, просто чуть-чуть попозже сделаю. Тут нужно просто редактор уже открывать и чуть повнимательнее вносить фикс, чтобы не убить всё.

ketchoop committed:
Add cache_dir var. Little refactoring

shuhrat commented on npm:
```diff
@@ -38,6 +39,24 @@ fi
 
 echo "==> Running npm wrapper script with node_modules caching"
 
+#Если существует переменная времени жизни кэша, то ищем просроченные кэши
```
пробел после `#`, может описание перенесем выше?

ketchoop replied:
Окей, сделаю.

shuhrat commented on npm:
```diff
@@ -38,6 +39,24 @@ fi
+if [[ -n ${NPM_CACHE_EXPIRE_DAYS} ]]; then
+  echo "Looking for expired cache"
+
+  CACHE_DIR="$HOME/.cache/npm-inject"
```
2 раза объявляем  `CACHE_DIR="$HOME/.cache/npm-inject"`

shuhrat commented on npm:
```diff
@@ -38,6 +39,24 @@ fi
+  echo "Looking for expired cache"
+
+  CACHE_DIR="$HOME/.cache/npm-inject"
+  EXPIRED_DIRS=$(find $CACHE_DIR -maxdepth 1 -mtime +$NPM_CACHE_EXPIRE_DAYS ! -path $CACHE_DIR -type d)
```
думаю `! -path $CACHE_DIR` не нужен

ketchoop replied:
@shuhrat Тогда, если каталог кэша(CACHE_DIR) будет "просрочен", то удалится и он, а вместе с ним и все кэши. Или я ошибаюсь?

ketchoop replied:
@shuhrat Протестил. Действительно удаляет, если возраст CACHE_DIR больше срока жизни кэшей, даже если внутри кэши не увляются устаревшими(что логично).

shuhrat commented on npm:
```diff
@@ -49,15 +68,15 @@ PACKAGES_HASH=$(openssl sha1 $PACKAGE_DEPS | awk '{print $2}')
 
 NODE_MODULES="$PWD/node_modules"
-CACHE_DIR="$HOME/.cache/npm-inject/$PACKAGES_HASH"
+PACKAGES_CACHE_DIR="$CACHE_DIR/$PACKAGES_HASH"
```
👍

ketchoop committed:
Comment fixes

Move description to the top and change it.
Add missing space after 'sharp'

ketchoop committed:
Delete redundant var

shuhrat commented:
Спасибо за пулл-реквест!

shuhrat commented:
https://yandex.ru/jobs/vacancies/dev/inter_infrastructure_dev/
@ketchoop может тебе будет интересно =)
